openapi/controller/main.py

```diff
@@ -0,0 +1,183 @@
+# -*- coding: utf-8 -*-
+import functools
+import hashlib
+import logging
+import os
+from ast import literal_eval
+import base64
+
+import odoo
+from odoo import http, SUPERUSER_ID, models
+from odoo.http import request, OpenERPSession
+from odoo.modules.registry import RegistryManager
+from psycopg2.extensions import ISOLATION_LEVEL_READ_COMMITTED
+
+from .pinguin import *
+
+_logger = logging.getLogger(__name__)
+
+#################################################################
+# Odoo REST API                                                 #
+#  Version 1                                                    #
+# --------------------------------------------------------------#
+# The current api version is considered stable, although        #
+# the exposed models and methods change as they are configured  #
+# on the database level. Only if significant changes in the api #
+# generation logic should be implemented in the future          #
+# a version bump should be considered.                          #
+#################################################################
+
+API_ENDPOINT = '/api'
+API_ENDPOINT_V1 = '/v1'
+# API_ENDPOINT_V2 = '/2'
+
+# We patch the route decorator in pinguin.py
+# with authentication and DB inference logic.
+# We also check if the model is installed in the database.
+# This keeps the code below minial and readable.
+http.route = pinguin_route
+
+
+class ApiV1Controller(http.Controller):
+    """ Implements the REST API V1 endpoint.
+    .. methods:
+
+        CRUD Methods:
+        - `POST     ../<model>`               -> `CreateOne`
+        - `PUT      ../<model>/<id>`          -> `UpdateOne`
+        - `GET      ../<model>`               -> `ReadMulti`
+        - `GET      ../<model>/<id>`          -> `ReadOne`
+        - `DELETE   ../<model>/<id>`          -> `UnlinkOne`
+
+        Auxiliary Methods:
+        - `PATCH    ../<model>/<id>/<method>`               -> `Call Method on Singleton Record`
+        - `PATCH    ../<model>/<method>`                    -> `Call Method on RecordSet`
+        - `GET      ../report/pdf/<report_external_id>`     -> `Get Report as PDF`
+        - `GET      ../report/html/<report_external_id>`    -> `Get Report as HTML`
+    """
+
+    _api_endpoint = API_ENDPOINT + API_ENDPOINT_V1
+    _api_endpoint = _api_endpoint + '/<namespace>'
+    # CreateOne # ReadMulti
+    _api_endpoint_model = _api_endpoint + '/<model>'
+    # ReadOne # UpdateOne # UnlikOne
+    _api_endpoint_model_id = _api_endpoint + '/<model>/<id>'
+    # Call Methods
+    _api_endpoint_model_id_method = _api_endpoint + '/<model>/<id>/<method>'  # on Singleton Record
+    _api_endpoint_model_method = _api_endpoint + '/<model>/<method>'  # on RecordSet
+    # Get Reports
+    _api_report_pdf = _api_endpoint + '/report/pdf/<report_external_id>'  # as PDF
+    _api_report_html = _api_endpoint + '/report/html/<report_external_id>'  # as HTML
+
+
+    ##################
+    ## CRUD Methods ##
+    ##################
```

Comment:
For me it's a bit blow up the code. It could be done in couple controllers:
```
@http.route([
'api/v<int:version>/<model>',
'api/v<int:version>/<model>/<int:id>',
], methods=['GET', 'POST', 'PUT', 'DELETE', 'PATCH'], 
type='http', auth='none', csrf=False)
def api_model(self, version, model, id=None):
```

Then either check methods and other
```
    if request.httprequest.method == 'GET':
          ....
```
or pass everything to pinguin

Reply:
It's on purpose, it should not do anything. I try to avoid any logic here, it should just forward to pinguin, but leave a clear and readable interface description...

One endpoint one method...